python/graphstorm/graphpartitioning/metadata_schema.py

```diff
@@ -0,0 +1,799 @@
+    in this class is
+    >>> ntypes = Metadata().ntypes # To retrieve node types from the graph
+    >>> # To get the id <-> ntype map
+    >>> # id_ntype_map will be {0: ``ntype0``, 1: ``ntype1``, ...}
+    >>> id_ntype_map = Metadata().id_ntype_map
+    """
+
+    _instance = None
+
+    def __new__(cls):
+        if cls._instance is None:
+            cls._instance = super(MetadataSchema, cls).__new__(cls)
+        return cls._instance
+
+    def __init__(self):
+        if self._instance is not None:
+            return
+
+        self._basedir = None
+        self._metadata_path = None
+        self._data = None
+        self._global_nid_offsets = None
+        self._global_eid_offsets = None
+        self._ntypes = None
+        self._ntype_id_map = None
+        self._id_ntype_map = None
+        self._etype_id_map = None
+        self._id_etype_map = None
+        self._etypes = None
+        self._ntype_features = None
+        self._etype_features = None
+        self._ntype_feature_files = None
+        self._etype_feature_files = None
+        self._etype_files = None
+
+    def _load_json(self, file_name, base_dir):
+        """Helpper function to read the metadata schema file of the input
+        graph dataset. ``file_name`` input argument can be either in absolute
+        or relative path formats. ``base_dir`` is used as the root directory
+        for the input graph dataset where all the files (of the dataset) are
+        expected to be located on the disk.
+
+        Argument
+        --------
+        file_name: string
+            Filename of the input metadata schema file. This can be either
+            in absolute or relative path formats. If in relative format,
+            ``base_dir`` is used as the root directory.
+        base_dir: string
+            Base directory, if provided, is used as the root directory for
+            the input graph dataset
+        """
+        if file_name is None or not isinstance(file_name, str):
+            raise ValueError(f"File path is not valid. = {file_name}")
+
+        self._basedir = None
+        if base_dir is not None and os.path.isdir(base_dir):
+            if not os.path.isfile(os.path.join(base_dir, file_name)):
+                raise ValueError(
+                    f"Metadata file does not exist. {base_dir}, {file_name}"
+                )
+            self._metadata_path = os.path.join(
+                os.path.join(base_dir, file_name)
+            )
+            self._basedir = base_dir
+        else:
+            if not os.path.isfile(file_name):
+                raise ValueError(f"Metadata file does not exist. {file_name}")
+            self._metadata_path = file_name
+
+        if self._basedir is None:
+            self._basedir = os.getcwd()
+
+        with open(self._metadata_path, "r", encoding="utf-8") as handle:
+            self._data = json.load(handle)
+
+    def _check_ifexists(self, data_files, src_type):
```

Comment:
In general it's hard to see the value of this function, just let Python error out if a file doesn't exist.

Reply:
The main reason for this function is to ensure that the files are present, as per their suggested locations in the metadata file. This has to do with the user experience of the application. 

Imagine a use case where the pipeline is reading thousands of files. For instance, webgraph has about 700+ files for edges. And if the very last file is miss-spelled or could not be located for some reason. In this case in the current implementation the user has to wait for hours to run into a FileNotFoundError, an avoidable capex which is proactively being saved in this instance. 

Either a check is to made to ensure the presence of the file or let the system handle the case when it encountered such a thing. A decision was in the favor of proactively checking the presence of these files.